odl/operator/fn_ops.py

```diff
@@ -0,0 +1,454 @@
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# ODL is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with ODL.  If not, see <http://www.gnu.org/licenses/>.
+#
+#"""Default operators defined on any (reasonable) space."""
+
+# Imports for common Python 2/3 codebase
+from __future__ import print_function, division, absolute_import
+from future import standard_library
+standard_library.install_aliases()
+from builtins import super
+
+from odl.operator.operator import Operator
+from odl.space import fn
+from odl.set import LinearSpace, Field
+from odl.discr import DiscreteLp
+from odl.space.base_ntuples import FnBase
+
+import numpy as np
+
+__all__ = ('SamplingOperator', 'ZerofillingOperator', 
+           'VectoriseOperator', 'EmbeddingOperator')
+
+class SamplingOperator(Operator):
+
+    """Operator that samples coefficients.
+    The operator is defined by 
+        ``SamplingOperator(f, indices) == cell_volume * f[indices]``
+    """
+
+    def __init__(self, domain, indices, weighting=None):
+        """Initialize a new instance.
+
+        Parameters
+        ----------
+        domain : `FnBase` or `DiscreteLp`
+            Set of elements on which this operator acts.
+        indices : list
+            List of indices where to sample the function
+        weighting : {None, 'cell_volume'}
+            weighting scheme
+
+        Examples
+        --------
+        >>> X = odl.uniform_discr(min_pt=[0, 0], max_pt=[1, 1], shape=[2, 2])
+        >>> x = X.element(range(X.size))
+        >>> ind = [[0, 1], [1, 1]]
+        >>> A = odl.SamplingOperator(X, ind)
+        >>> A(x)
+        rn(2).element([1.0, 3.0])
+        >>> A(x)  # Out-of-place
+        rn(2).element([1.0, 3.0])
+        """
+        if not isinstance(domain, (FnBase, DiscreteLp)):
+            raise TypeError('`space` {!r} not a `FnBase` or `DiscreteLp` '
+                            'instance'.format(domain))
+
+        self.__indices = np.asarray(indices, dtype=int)
+        # Converts a single sequence to an array of integers
+        # and a sequence of arrays to a vertically stacked array
+        if self.indices.ndim > 1:
+            # Strides = increment in linear indices per axis
+            strides = np.concatenate((np.cumprod(domain.shape[1:])[::-1], [1]))
+            self.__indices_flat = np.sum(self.indices * strides[:, None], axis=0)
+        else:
+            self.__indices_flat = self.indices
+
+        self.__weighting = weighting        
+        if self.weighting is None:
+            self.__weights = 1
+        elif self.weighting == 'cell_volume':
+            self.__weights = getattr(domain, 'cell_volume', 1.0)
```

Comment:
I could do this, but then the call is more complicated as we need the if statement there then. I prefer it to be this way to be honest.